Payroll all rubrics

lomamech opened the pull request:
refactor(Payroll) 
-Remove all old payroll table for bhima 1.x (Tax, Rubric, Cotisation)
- Implement API and client side for Payroll Rubric
- Implement Integration test for Payroll Rubrics
- Implement end to end test for Payroll Rubrics

reference #2292

lomamech committed:
refactor(Payroll)
-Remove all old payroll table for bhima 1.x (Tax, Rubric, Cotisation)

lomamech committed:
test(Payroll Rubrics)
- Implement Integration test for Payroll Rubrics
- Implement end to end test for Payroll Rubrics

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -314,6 +320,7 @@
             "FINANCIAL_DETAIL": "Current balance, the posting journal is included:",
             "FINANCIAL_STATUS": "Financial Status",
             "FIRST_NAME": "First Name",
+            "FOUR_ACCOUNT": "Third parties accounts",
```
Instead of calling this "Four Account", can we say "debtor_account" or something like that?   That way it isn't hard-coded OHADA.  "Four account" doesn't mean anything in English.

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -454,6 +464,7 @@
             "SEX": "Sex",
             "SHOW_SIMPLE_PREVIEW": "Show Simple Preview",
             "SIGNATURE": "Signature",
+            "SIX_ACCOUNT": "Costs Accounts",
```
Instead of "Six Account", can we call this "Expense Account"?  That way, we will be able to explain it to people who are not familiar with OHADA.  Six accounts are just expense accounts.

jniles commented on server/controllers/finance/accounts/index.js:
```diff
@@ -117,6 +118,8 @@ function remove(req, res, next) {
  * GET /accounts
  */
 function list(req, res, next) {
+  const filters = new FilterParser(req.query, { tableAlias : 'a' });
```
Nice :+1:  Contributes to #2240.

jniles commented on client/src/js/components/bhAccountSelect.js:
```diff
@@ -36,6 +37,9 @@ function AccountSelectController(Accounts, AppCache, $timeout, bhConstants, $sco
     $ctrl.TITLE_ACCOUNT_ID = bhConstants.accounts.TITLE;
 
+    // To filter accounts by class of accounts
+    $ctrl.classe = $ctrl.classe || '';
```
I really think this is a duplication of `account_type_id`.  Let's hold off on this dependency until we are certain we want to use `classe` instead of `account_type`.  After all, users can view and modify the `type_id` on the account page, but they cannot modify the `classe`.

jniles commented on client/src/modules/payroll/rubrics/rubrics.js:
```diff
@@ -0,0 +1,110 @@
+  function loadRubrics() {
+    vm.loading = true;
+
+    Rubrics.read(null, { detailed : 1 })
```
Does rubrics have a detailed option?

jniles commented on client/src/modules/payroll/rubrics/rubrics.js:
```diff
@@ -0,0 +1,110 @@
+
+  // create a new Rubric
+  function createRubric() {
+    $state.go('rubrics.create');
```
These state links should just be directives using `ui-sref`.  It will be easier to migrate to future versions of angular that way, I think.

jniles commented on client/src/modules/payroll/rubrics/templates/ipr.tmpl.html:
```diff
@@ -0,0 +1,5 @@
+<div class="ui-grid-cell-contents text-left">
+  <div ng-if="row.entity.is_ipr">
+  	<i class="fa fa-check-square"></i>
```
In my experience at HEV, users don't really know what the icons are.  Plus they have resizing issues in the grid (many users with poor eyesight increase the screen size and the icons then look very distorted).  It would be better to use "oui/non" or "true/false" for boolean values, I think.

jniles commented on test/end-to-end/rubrics/rubrics.spec.js:
```diff
@@ -0,0 +1,48 @@
+    abbr  : 'CoSynd',
+    is_employee : 1,
+    is_percent : 1,
+    third_party_account_id : '40111002',
```
Could you include a comment with the account name here?  Like:
```
 third_party_account_id : '40111002', // SNEL
```
or something?  That will help future developers know what account's balance is changing.

jniles commented on test/integration/rubrics.js:
```diff
@@ -0,0 +1,90 @@
+  });
+
+  it('PUT /RUBRICS  should update an existing Rubric ', function () {
+    return agent.put('/rubrics/' + cotisation.id)
```
Could you make these tests respect eslint?

jniles commented on server/models/schema.sql:
```diff
@@ -1687,43 +1542,6 @@ CREATE TABLE `supplier` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
-
-DROP TABLE IF EXISTS `tax`;
```
Wee!!

jniles commented on client/src/modules/payroll/rubrics/modals/rubric.modal.html:
```diff
@@ -0,0 +1,172 @@
+    </bh-account-select>
+
+    <bh-account-select
+      id="costs_account_id"
```
A better name is "Expense Account".  Tax is an expense, for example.  The account type it will use is an expense account as well.

jniles commented on client/src/modules/payroll/rubrics/modals/rubric.modal.js:
```diff
@@ -0,0 +1,77 @@
+
+  vm.setDiscount = function setDiscount(value) {
+    vm.setting = true;
+    vm.rubric.is_discount = value === 1 ? true : false;
```
Can you use `ng-true-value` and `ng-false-value` to do this automatically in the view?

jniles reviewed (commented):
@lomamech this represents a large amount of work.   I've tested this and the user experience of creating a rubric is pretty good!  You might consider using `inline` checkboxes and radio buttons instead of the vertical ones to save space.

I'm not certain we should go towards using account `classe` yet.  We've had it for historical reasons, but it really is better to use account types.  We have pretty fine-grained account types now that roughly correspond with the `classe` in OHADA (https://github.com/IMA-WorldHealth/bhima-2.X/issues/1607).

Filtering by account type is a good idea!  We should implement it (https://github.com/IMA-WorldHealth/bhima-2.X/issues/2364).

In general, we probably shouldn't use icons as the only piece of text in a grid column.  They look terrible when printed out, and they don't convey a lot of information to users who cannot recognize what they are.  For true/false values, let's just say that.

I'm not very experienced with the rules for apply taxes and payrolls, so I recommend you run this by @mbayopanda or @DedrickEnc who previously worked on the payroll forms in 1.X to see what they think.

Overall, this is a heap of work and pretty good code!  Nice!  :+1:

lomamech committed:
Resolve Conflict and Rebase with Branch Master

lomamech committed:
refactor(Rubrics Payroll)
- Using class radio-inline for radio buttons

lomamech committed:
refactor(Payroll All R

lomamech committed:
Refactor(Payroll All Rubrics)
- Use Expense Account instead Six Account
- Use Debtor Accoubt instead Third parties Accounts
- Use ng-maxlength in abbr field for limite Abbreviation
- Refactor Is Discount use value D for Discount and A for addition for
  simplify the management
- Prevent percent value to be superior to 100
- Prevent to be inferior to Zero
- Use UI-SREF for display rubric create Modal
- Update ui-grid rubrics properties and column element
- Remplace all icon in Rubrics UI-grid and Use Text

lomamech committed:
Sanitaze Code

lomamech committed:
perf(Payroll Rubrics)
- Improve integration test of rubric use concat instead signe Plus

jniles reviewed (approved):
LGTM.  A lot of work went into this!

bors r+
